packages/core/src/storages/request_provider.ts

```diff
@@ -567,6 +567,30 @@ export abstract class RequestProvider implements IStorage {
      * the function might occasionally return a false negative,
      * but it will never return a false positive.
      */
+    /**
+     * TODO Kuba:
+     * I would move this to request_queue_v2.ts file as it is related to the new API.
+     * Let's clean it up. We would probably need some new API endpoints to make this work.
+     * But for now, we can use the listHead and getRequest as hacky solution.
+     * Because the listHead list all requests that are not handled, event the locked ones. And we can use getRequest to check if the request is locked.
+     *
+     * There are basically two paths here:
+     * 1. There are no other clients touching the queue(hadMultipleClient === false), so we can trust the to queueHeadIds cache.
+     * - If the queueHeadIds is not empty -> happy path we have some requests to process. -> false
+     * - If the queueHeadIds is empty
+     *     -> we need to ensure that the queue is really empty(ensureHeadIsNonEmpty).
+     *        -> Check again if the queueHeadIds is not empty -> false
+     *
+     * 2. There are other clients touching the queue(hadMultipleClient === true), so we need to count that other client can change the head.
+     * - If the queueHeadIds is not empty -> happy path we have some requests to process. -> false
+     * - If the queueHeadIds is empty
+     *     -> we need to ensure that the queue is really empty(ensureHeadIsNonEmpty).
+     *     -> we need to check if other clients still processing the requests (using listHead and getRequest on the first request)
```

Comment:
What does the `getRequest` accomplish here? We already established that our queue head is empty (line above), and if `listHead` returns a non-empty list, it means that a) some request arrived in the meantime or b) some other client is still working. Both means we are not finished, doesn't it?

Reply:
Pardon me, I'm stuck at "type of request" - what should I be looking for? If we're just interested in lock status, isn't it already certain that the request is locked by another client if `ensureHeadIsNonEmpty` returns nothing but `listHead` still returns requests?

Reply:
Yeah, the listHead returns even locked requests plus pending. This way you can list all pending and lock requests for the queue, and then check if other clients lock these. 
But as I write this I feel strange as it does not make sense 😄 . Anyway we should fix this return just pending requests without locked and introduce new API to list just lock reauests by client.

Reply:
But... if I got nothing from `listAndLockHead`, there should be no pending requests..? Right? Or if there are (added between `listAndLockHead` and the subsequent `listHead`), isn't that what that grace period is for?

Reply:
> But... if I got nothing from listAndLockHead, there should be no pending requests..?

Yes

> Or if there are (added between listAndLockHead and the subsequent listHead), isn't that what that grace period is for?

I think the listHead is used there just get locked request. But true that it can be used this way, but you can do the same with listAndLockHead and the subsequent listAndLockHead.

Reply:
Okay, so what you suggest in the comment is:

1. try `listAndLockHead`
2. if it returns nothing, try `listHead`
3. if there's something, it might be a) a request that arrived between 1. and 2. or b) a request locked by another client, and we detect that using `getRequest`, but this whole shtick is just so that we have more information to log

Do I understand this right?

Reply:
Yep, can sound strange as you write it 😄 , I think the user should have an idea that he is paying to compute units to wait for another run to finish its job or not.